packages/base-driver/lib/basedriver/capabilities.js

```diff
@@ -10,17 +10,21 @@ const APPIUM_VENDOR_PREFIX = 'appium:';
 const APPIUM_OPTS_CAP = 'options';
 const PREFIXED_APPIUM_OPTS_CAP = `${APPIUM_VENDOR_PREFIX}${APPIUM_OPTS_CAP}`;
 
-// Takes primary caps object and merges it into a secondary caps object.
-// (see https://www.w3.org/TR/webdriver/#dfn-merging-capabilities)
 /**
- * @param {Capabilities} [primary]
- * @param {Capabilities} [secondary]
- * @returns {Capabilities}
+ * Takes primary caps object and merges it into a secondary caps object.
+ * @template {Constraints} [T={}]
+ * @template {Constraints} [U={}]
+ * @param {Capabilities<T>} [primary]
+ * @param {Capabilities<U>} [secondary]
+ * @returns {import('type-fest').Merge<Capabilities<T>, Capabilities<U>>}
+ * @see https://www.w3.org/TR/webdriver/#dfn-merging-capabilities)
  */
 function mergeCaps(primary = {}, secondary = {}) {
-  let result = Object.assign({}, primary);
+  let result = /** @type {import('type-fest').Merge<Capabilities<T>, Capabilities<U>>} */ ({
```

Comment:
so, I thought about this, but then I'm wondering if we should.

capabilities (the values thereof), afaict, cannot be plain objects.  they _can_ be arrays.  so what are we doing when we're merging arrays here?  the code (before I changed it) didn't handle this case, and simply overwrites the `primary` array with the `secondary` array.

we're not mutating the arrays to my knowledge (at least not since I changed that fn not to mutate), so I _think_ we should be in the clear, but it begs the question: are we doing the right thing w/r/t merging caps?  @jlipps

Reply:
hmm... capabilities can be objects, but I don't see any way to declare that in constraints.